engine/table/src/main/java/io/deephaven/engine/table/impl/naturaljoin/typed/incopen/gen/IncrementalNaturalJoinHasherChar.java

```diff
@@ -60,17 +60,26 @@ protected void buildFromLeftSide(RowSequence rowSequence, Chunk[] sourceKeyChunk
             final int hash = hash(k0);
             final int firstTableLocation = hashToTableLocation(hash);
             int tableLocation = firstTableLocation;
+            int firstDeletedLocation = -1;
             MAIN_SEARCH: while (true) {
                 long rightRowKeyForState = mainRightRowKey.getUnsafe(tableLocation);
-                if (rightRowKeyForState == EMPTY_RIGHT_STATE) {
-                    numEntries++;
+                if (firstDeletedLocation < 0 && isStateDeleted(rightRowKeyForState)) {
+                    firstDeletedLocation = tableLocation;
+                }
+                if (isStateEmpty(rightRowKeyForState)) {
+                    if (firstDeletedLocation >= 0) {
+                        tableLocation = firstDeletedLocation;
+                    } else {
+                        numEntries++;
+                    }
+                    liveEntries++;
                     mainKeySource0.set(tableLocation, k0);
                     mainLeftRowSet.set(tableLocation, RowSetFactory.fromKeys(rowKeyChunk.get(chunkPosition)));
                     mainRightRowKey.set(tableLocation, RowSet.NULL_ROW_KEY);
                     mainModifiedTrackerCookieSource.set(tableLocation, -1L);
                     break;
-                } else if (eq(mainKeySource0.getUnsafe(tableLocation), k0)) {
-                    if (rightRowKeyForState < RowSet.NULL_ROW_KEY) {
+                } else if (!isStateDeleted(rightRowKeyForState) && eq(mainKeySource0.getUnsafe(tableLocation), k0)) {
+                    if (rightRowKeyForState <= FIRST_DUPLICATE) {
```

Comment:
I think we should check for key-equality unconditionally. If we've found ourselves, but we're deleted, we're able to cease probing and insert at `firstDeletedLocation`.

Reply:
We can use this to stop probes early, too. That said, it does mean we read keys for deleted slots. Maybe there are workloads when that is worse?